modules/EvseManager/IECStateMachine.cpp

```diff
@@ -0,0 +1,451 @@
+    case CPEvent::MREC_2_GroundFailure:
+        return "MREC_2_GroundFailure";
+    case CPEvent::MREC_3_HighTemperature:
+        return "MREC_3_HighTemperature";
+    case CPEvent::MREC_4_OverCurrentFailure:
+        return "MREC_4_OverCurrentFailure";
+    case CPEvent::MREC_5_OverVoltage:
+        return "MREC_5_OverVoltage";
+    case CPEvent::MREC_6_UnderVoltage:
+        return "MREC_6_UnderVoltage";
+    case CPEvent::MREC_8_EmergencyStop:
+        return "MREC_8_EmergencyStop";
+    case CPEvent::MREC_15_PowerLoss:
+        return "MREC_15_PowerLoss";
+    case CPEvent::MREC_17_EVSEContactorFault:
+        return "MREC_17_EVSEContactorFault";
+    case CPEvent::MREC_18_CableOverTempDerate:
+        return "MREC_18_CableOverTempDerate";
+    case CPEvent::MREC_19_CableOverTempStop:
+        return "MREC_19_CableOverTempStop";
+    case CPEvent::MREC_20_PartialInsertion:
+        return "MREC_20_PartialInsertion";
+    case CPEvent::MREC_23_ProximityFault:
+        return "MREC_23_ProximityFault";
+    case CPEvent::MREC_24_ConnectorVoltageHigh:
+        return "MREC_24_ConnectorVoltageHigh";
+    case CPEvent::MREC_25_BrokenLatch:
+        return "MREC_25_BrokenLatch";
+    case CPEvent::MREC_26_CutCable:
+        return "MREC_26_CutCable";
+    }
+    throw std::out_of_range("No known string conversion for provided enum of type CPEvent");
+}
+
+IECStateMachine::IECStateMachine(const std::unique_ptr<evse_board_supportIntf>& r_bsp) : r_bsp(r_bsp) {
+    // feed the state machine whenever the timer expires
+    timeout_state_c1.signal_reached.connect(&IECStateMachine::feed_state_machine, this);
+
+    // Subscribe to bsp driver to receive BspEvents from the hardware
+    r_bsp->subscribe_event([this](const types::board_support_common::BspEvent event) {
+        // feed into state machine
+        process_bsp_event(event);
+    });
+}
+
+void IECStateMachine::process_bsp_event(types::board_support_common::BspEvent bsp_event) {
+
+    auto event_or_error = from_bsp_event(bsp_event.event);
+
+    // If it was a CP change: feed state machine, else forward error
+    if (std::holds_alternative<RawCPState>(event_or_error)) {
+        {
+            std::lock_guard l(state_mutex);
+            cp_state = std::get<RawCPState>(event_or_error);
+        }
+        feed_state_machine();
+    } else {
+        signal_event(std::get<CPEvent>(event_or_error));
+    }
+}
+
+void IECStateMachine::feed_state_machine() {
+    auto events = state_machine();
+
+    // Process all events
+    while (not events.empty()) {
+        signal_event(events.front());
+        events.pop();
+    }
+}
+
+// Main IEC state machine. Needs to be called whenever:
+// - CP state changes (both events from hardware as well as duty cycle changes)
+// - Allow power on changes
+// - The C1 6s timer expires
+std::queue<CPEvent> IECStateMachine::state_machine() {
+
+    std::queue<CPEvent> events;
+    std::lock_guard lock(state_mutex);
+
+    switch (cp_state) {
```

Comment:
would it make sense to get the copy of the cp-state here and just to release the mutex?

Reply:
The mutex protects more then just the cp_state variable. I renamed it to state_machine_mutex to make that point more clear.